File: zy1177/backend/app/algorithms/pathfinding.py
import heapq
import math
from typing import List, Tuple, Dict, Optional
from dataclasses import dataclass


@dataclass
class Node:
    x: int
    y: int
    g: float = 0
    h: float = 0
    f: float = 0
    parent: Optional['Node'] = None
    
    def __lt__(self, other):
        return self.f < other.f
    
    def __eq__(self, other):
        return self.x == other.x and self.y == other.y
    
    def __hash__(self):
        return hash((self.x, self.y))


class PathfindingAlgorithm:
    def __init__(self, grid: List[List[int]], grid_size: float = 1.0):
        self.grid = grid
        self.grid_size = grid_size
        self.width = len(grid[0]) if grid else 0
        self.height = len(grid) if grid else 0
    
    def heuristic(self, a: Tuple[int, int], b: Tuple[int, int]) -> float:
        dx = abs(a[0] - b[0])
        dy = abs(a[1] - b[1])
        return math.sqrt(dx * dx + dy * dy)
    
    def get_neighbors(self, node: Node, allow_diagonal: bool = True) -> List[Tuple[int, int]]:
        directions = [
            (0, -1), (0, 1), (-1, 0), (1, 0),
        ]
        if allow_diagonal:
            directions.extend([(-1, -1), (-1, 1), (1, -1), (1, 1)])
        
        neighbors = []
        for dx, dy in directions:
            nx = node.x + dx
            ny = node.y + dy
            if 0 <= nx < self.width and 0 <= ny < self.height:
                if self.grid[ny][nx] == 0:
                    neighbors.append((nx, ny))
        return neighbors
    
    def a_star(
        self, 
        start: Tuple[float, float], 
        goal: Tuple[float, float],
        allow_diagonal: bool = True,
        robot_radius: float = 0.5
    ) -> List[Tuple[float, float]]:
        start_grid = (
            int(start[0] / self.grid_size),
            int(start[1] / self.grid_size)
        )
        goal_grid = (
            int(goal[0] / self.grid_size),
            int(goal[1] / self.grid_size)
        )
        
        if start_grid[0] < 0 or start_grid[0] >= self.width or \
           start_grid[1] < 0 or start_grid[1] >= self.height:
            return []
        
        if goal_grid[0] < 0 or goal_grid[0] >= self.width or \
           goal_grid[1] < 0 or goal_grid[1] >= self.height:
            return []
        
        if self.grid[start_grid[1]][start_grid[0]] != 0:
            return []
        
        if self.grid[goal_grid[1]][goal_grid[0]] != 0:
            return []
        
        open_set = []
        closed_set = set()
        
        start_node = Node(start_grid[0], start_grid[1])
        goal_node = Node(goal_grid[0], goal_grid[1])
        
        start_node.g = 0
        start_node.h = self.heuristic(start_grid, goal_grid)
        start_node.f = start_node.g + start_node.h
        
        heapq.heappush(open_set, start_node)
        
        while open_set:
            current_node = heapq.heappop(open_set)
            
            if current_node == goal_node:
                path = []
                while current_node:
                    path.append((
                        current_node.x * self.grid_size + self.grid_size / 2,
                        current_node.y * self.grid_size + self.grid_size / 2
                    ))
                    current_node = current_node.parent
                return path[::-1]
            
            closed_set.add(current_node)
            
            for neighbor_pos in self.get_neighbors(current_node, allow_diagonal):
                neighbor = Node(neighbor_pos[0], neighbor_pos[1])
                
                if neighbor in closed_set:
                    continue
                
                dx = neighbor.x - current_node.x
                dy = neighbor.y - current_node.y
                move_cost = math.sqrt(dx * dx + dy * dy) * self.grid_size
                
                tentative_g = current_node.g + move_cost
                
                if neighbor not in open_set:
                    neighbor.g = tentative_g
                    neighbor.h = self.heuristic(neighbor_pos, goal_grid)
                    neighbor.f = neighbor.g + neighbor.h
                    neighbor.parent = current_node
                    heapq.heappush(open_set, neighbor)
                elif tentative_g < neighbor.g:
                    neighbor.g = tentative_g
                    neighbor.f = neighbor.g + neighbor.h
                    neighbor.parent = current_node
        
        return []
    
    def smooth_path(self, path: List[Tuple[float, float]]) -> List[Tuple[float, float]]:
        if len(path) <= 2:
            return path
        
        smoothed = [path[0]]
        current_index = 0
        
        while current_index < len(path) - 1:
            farthest_reachable = current_index + 1
            
            for j in range(len(path) - 1, current_index, -1):
                if self._is_line_clear(path[current_index], path[j]):
                    farthest_reachable = j
                    break
            
            if farthest_reachable > current_index + 1:
                smoothed.append(path[farthest_reachable])
                current_index = farthest_reachable
            else:
                smoothed.append(path[current_index + 1])
                current_index += 1
        
        return smoothed
    
    def _is_line_clear(self, start: Tuple[float, float], end: Tuple[float, float]) -> bool:
        start_x, start_y = start
        end_x, end_y = end
        
        steps = max(
            int(abs(end_x - start_x) / self.grid_size) + 1,
            int(abs(end_y - start_y) / self.grid_size) + 1
        )
        
        for i in range(steps + 1):
            t = i / steps
            x = start_x + (end_x - start_x) * t
            y = start_y + (end_y - start_y) * t
            
            grid_x = int(x / self.grid_size)
            grid_y = int(y / self.grid_size)
            
            if grid_x < 0 or grid_x >= self.width or \
               grid_y < 0 or grid_y >= self.height:
                return False
            
            if self.grid[grid_y][grid_x] != 0:
                return False
        
        return True
    
    def calculate_path_length(self, path: List[Tuple[float, float]]) -> float:
        if len(path) < 2:
            return 0
        
        total = 0
        for i in range(1, len(path)):
            dx = path[i][0] - path[i-1][0]
            dy = path[i][1] - path[i-1][1]
            total += math.sqrt(dx * dx + dy * dy)
        return total

File: zy71102/src/simulation/pathfinding.ts
import { Point, Wall, ClosedArea, Gate, GRID_SIZE } from './types';

interface GridNode {
  x: number;
  y: number;
  g: number;
  h: number;
  f: number;
  parent: GridNode | null;
  walkable: boolean;
}

export class PathFinder {
  private gridWidth: number;
  private gridHeight: number;
  private walls: Wall[];
  private closedAreas: ClosedArea[];
  private gates: Gate[];

  constructor(
    stationWidth: number,
    stationHeight: number,
    walls: Wall[],
    closedAreas: ClosedArea[],
    gates: Gate[]
  ) {
    this.gridWidth = Math.ceil(stationWidth / GRID_SIZE);
    this.gridHeight = Math.ceil(stationHeight / GRID_SIZE);
    this.walls = walls;
    this.closedAreas = closedAreas;
    this.gates = gates;
  }

  private worldToGrid(worldX: number, worldY: number): { x: number; y: number } {
    return {
      x: Math.floor(worldX / GRID_SIZE),
      y: Math.floor(worldY / GRID_SIZE)
    };
  }

  private gridToWorld(gridX: number, gridY: number): Point {
    return {
      x: (gridX + 0.5) * GRID_SIZE,
      y: (gridY + 0.5) * GRID_SIZE
    };
  }

  private isWalkable(gridX: number, gridY: number): boolean {
    if (gridX < 0 || gridX >= this.gridWidth || gridY < 0 || gridY >= this.gridHeight) {
      return false;
    }

    const worldPos = this.gridToWorld(gridX, gridY);

    for (const wall of this.walls) {
      if (
        worldPos.x >= wall.x &&
        worldPos.x <= wall.x + wall.width &&
        worldPos.y >= wall.y &&
        worldPos.y <= wall.y + wall.height
      ) {
        return false;
      }
    }

    for (const area of this.closedAreas) {
      if (
        worldPos.x >= area.x &&
        worldPos.x <= area.x + area.width &&
        worldPos.y >= area.y &&
        worldPos.y <= area.y + area.height
      ) {
        return false;
      }
    }

    for (const gate of this.gates) {
      if (gate.status === 'closed') {
        const gateWidth = 1.5;
        const gateHeight = 0.5;
        if (
          worldPos.x >= gate.x - gateWidth / 2 &&
          worldPos.x <= gate.x + gateWidth / 2 &&
          worldPos.y >= gate.y - gateHeight / 2 &&
          worldPos.y <= gate.y + gateHeight / 2
        ) {
          return false;
        }
      }
    }

    return true;
  }

  private heuristic(a: GridNode, b: GridNode): number {
    return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
  }

  private getNeighbors(node: GridNode, grid: GridNode[][]): GridNode[] {
    const neighbors: GridNode[] = [];
    const directions = [
      { dx: 0, dy: -1 },
      { dx: 1, dy: 0 },
      { dx: 0, dy: 1 },
      { dx: -1, dy: 0 },
      { dx: -1, dy: -1 },
      { dx: 1, dy: -1 },
      { dx: 1, dy: 1 },
      { dx: -1, dy: 1 }
    ];

    for (const dir of directions) {
      const nx = node.x + dir.dx;
      const ny = node.y + dir.dy;
      if (nx >= 0 && nx < this.gridWidth && ny >= 0 && ny < this.gridHeight) {
        if (grid[ny][nx].walkable) {
          if (dir.dx !== 0 && dir.dy !== 0) {
            if (!grid[node.y][node.x + dir.dx].walkable || !grid[node.y + dir.dy][node.x].walkable) {
              continue;
            }
          }
          neighbors.push(grid[ny][nx]);
        }
      }
    }

    return neighbors;
  }

  public findPath(startX: number, startY: number, endX: number, endY: number): Point[] {
    const startGrid = this.worldToGrid(startX, startY);
    const endGrid = this.worldToGrid(endX, endY);

    if (!this.isWalkable(endGrid.x, endGrid.y)) {
      return [];
    }

    const grid: GridNode[][] = [];
    for (let y = 0; y < this.gridHeight; y++) {
      grid[y] = [];
      for (let x = 0; x < this.gridWidth; x++) {
        grid[y][x] = {
          x,
          y,
          g: Infinity,
          h: 0,
          f: Infinity,
          parent: null,
          walkable: this.isWalkable(x, y)
        };
      }
    }

    const openList: GridNode[] = [];
    const closedSet = new Set<string>();

    const startNode = grid[startGrid.y][startGrid.x];
    const endNode = grid[endGrid.y][endGrid.x];

    if (!startNode.walkable) {
      return [];
    }

    startNode.g = 0;
    startNode.h = this.heuristic(startNode, endNode);
    startNode.f = startNode.h;
    openList.push(startNode);

    while (openList.length > 0) {
      openList.sort((a, b) => a.f - b.f);
      const current = openList.shift()!;

      if (current.x === endNode.x && current.y === endNode.y) {
        const path: Point[] = [];
        let node: GridNode | null = current;
        while (node) {
          path.unshift(this.gridToWorld(node.x, node.y));
          node = node.parent;
        }
        return this.smoothPath(path);
      }

      closedSet.add(`${current.x},${current.y}`);

      for (const neighbor of this.getNeighbors(current, grid)) {
        if (closedSet.has(`${neighbor.x},${neighbor.y}`)) {
          continue;
        }

        const tentativeG = current.g + (neighbor.x !== current.x && neighbor.y !== current.y ? 1.414 : 1);

        if (tentativeG < neighbor.g) {
          neighbor.parent = current;
          neighbor.g = tentativeG;
          neighbor.h = this.heuristic(neighbor, endNode);
          neighbor.f = neighbor.g + neighbor.h;

          if (!openList.includes(neighbor)) {
            openList.push(neighbor);
          }
        }
      }
    }

    return [];
  }

  private smoothPath(path: Point[]): Point[] {
    if (path.length < 3) return path;

    const smoothed: Point[] = [path[0]];
    let i = 0;

    while (i < path.length - 1) {
      for (let j = path.length - 1; j > i; j--) {
        if (this.hasLineOfSight(path[i], path[j])) {
          smoothed.push(path[j]);
          i = j;
          break;
        }
      }
    }

    return smoothed;
  }

  private hasLineOfSight(start: Point, end: Point): boolean {
    const dx = end.x - start.x;
    const dy = end.y - start.y;
    const steps = Math.max(Math.abs(dx), Math.abs(dy)) * 2;

    for (let i = 0; i <= steps; i++) {
      const t = i / steps;
      const x = start.x + dx * t;
      const y = start.y + dy * t;
      const grid = this.worldToGrid(x, y);
      if (!this.isWalkable(grid.x, grid.y)) {
        return false;
      }
    }

    return true;
  }
}
